### DISCUSSION TITLE
Cannot connect

### DISCUSSION THREAD

**Discussion by adsharma9** [winder/Universal-G-Code-Sender#2526] [Help]
I am trying Universal GCode sender for the first time with Arduino Uno.
I am running a Window10 machine. Arduino is 1.8.19
I have tried with UGS version 2.1.2 and  2.1.4. I get 
*** Connecting to jserialcomm://COM4:115200
*** Fetching device status
>>> ?
*** Fetching device status (2 of 3)...
>>> ?
*** Fetching device status (3 of 3)...
>>> ?
**** Connection closed ****
and with UGS version2.1.7
 i get same error after 10 tries.
I replaced the Arduino UNO with a new one. I checked the Arduino Uno using a CNC test program which runs the stepper motor one at a time this works. so the Arduino and the CNC shields are working
Please help.
Thanks Arvind

**Comment by jmonger1947** [comment]
I'm having the same problem with A Teensy4.1 board programed with GRBLHAL. UGS can connect to the board if I don't power cycle the board (basically unplug/replug it). If I unplug it I get the same message as adsharma9 AS long as I don't unplug it, I can even go to the Arduino serial monitor and connect which it won't do after the power cycle. A strange thing is I can go to TinyG part of UGS and connect with it then I can go to GRBL and it will connect correctly. I can disconnect and reconnect correctly as long as I don't unplug the board. So what's the difference between connecting with TinyG and GRBL and why does it continue to work after the initial connection with TinyG?

**Reply by breiler** [reply]
Please create a new issue. I need need the exact versions of everything that you are running (OS, UGS, grblHAL). I would also like to see the verbose log (right click in the console and choose to see detailed messages).

**Comment by aimeiz** [comment]
Hi.
I am trying to connect my machine running GRBL ESP on ESP32 GRBL version 1.3a.
UGS connect for a moment, displays some data taken from GRBL controller and then dosconnect with message: 
*** Could not detect the GRBL version
It is strange, since it gets and displays :
Grbl 1.3a ['$' for help]

Maybe Grbl 1.3a is not valid version for UGS Product Version: Universal Gcode Sender 20240704
Would You be kind making UGS less strict on version detection.
I am able to recompile my GRBL and set reported version to valid one, but I do not know what is valid.
Problem occurs only on TCP connection. On serial / USB. Bluetooth it works perfectly.

Best regards.

Marcin.

**Reply by breiler** [reply]
Exactly which firmware have you installed (link to the project page)?

Can you enable verbose logging in the console, right click and enable it. And then paste the complete log here.

UGS is now very accepting of the version string. It gets the version from $I command and not the welcome string. But there are some vendors that ships their controllers where they've stripped the information which has caused some problems.

**Comment by aimeiz** [comment]
Hi.
Than You for intimidate response.
Project taken from GitHub https://github.com/bdring/Grbl_Esp32
Customized for CNC controller I am using, but it only affect machine configuration file.

I choose GRBL ESP32 Firmware in UGS.
Console verbose mode seems to be default, but indicated as disabled. A bit strange.
When I enabled verbose mode I got only this:
Grbl 1.3a ['$' for help]

When disabled, this:
*** Connecting to tcp://192.168.1.153:23
Grbl 1.3a ['$' for help]
<Alarm|WPos:-22.999,-2.999,37.933,0.000,0.000|Bf:15,1200|FS:0,0|WCO:22.999,2.999,-37.933,0.000,0.000>
*** Fetching device status
>>> ?
<Alarm|WPos:-22.999,-2.999,37.933,0.000,0.000|Bf:15,1199|FS:0,0|Ov:100,100,100>
>>> 
ok
>>> $I
ok
*** Could not detect the GRBL version
**** Connection closed ****

If connected via COM port, result is as below:
*** Connecting to jserialcomm://COM3:115200
*** Fetching device status
>>> ?
<Alarm|WPos:-22.999,-2.999,37.933,0.000,0.000|Bf:15,127|FS:0,0>
ok
>>> 
ok
>>> $I
[VER:1.3a.20211103:]
[OPT:MPHBSW]
[MSG:Using machine:MKS TINYBEE]
[MSG:Mode=STA:SSID=Marcin_2.4G:Status=Connected:IP=192.168.1.153:MAC=D4-8A-FC-DF-D3-F8]
[MSG:No BT]
ok
>>> $$
$0 = 10    (Step pulse time, microseconds)
$1 = 254    (Step idle delay, milliseconds)
$2 = 0    (Step pulse invert, mask)
$3 = 5    (Step direction invert, mask)
$4 = 0    (Invert step enable pin, boolean)
$5 = 1    (Invert limit pins, boolean)
$6 = 1    (Invert probe pin, boolean)
$10 = 2    (Status report options, mask)
$11 = 0.010    (Junction deviation, millimeters)
$12 = 0.002    (Arc tolerance, millimeters)
$13 = 0    (Report in inches, boolean)
$20 = 1    (Soft limits enable, boolean)
$21 = 0    (Hard limits enable, boolean)
$22 = 1    (Homing cycle enable, boolean)
$23 = 3    (Homing direction invert, mask)
$24 = 250.000    (Homing locate feed rate, mm/min)
$25 = 1000.000    (Homing search seek rate, mm/min)
$26 = 250.000    (Homing switch debounce delay, milliseconds)
$27 = 3.000    (Homing switch pull-off distance, millimeters)
$30 = 10000.000    (Maximum spindle speed, RPM)
$31 = 0.000    (Minimum spindle speed, RPM)
$32 = 1    (Laser-mode enable, boolean)
$100 = 801.200    (X-axis travel resolution, step/mm)
$101 = 802.500    (Y-axis travel resolution, step/mm)
$102 = 802.500    (Z-axis travel resolution, step/mm)
$103 = 802.500   
$104 = 802.500   
$105 = 100.000   
$110 = 3000.000    (X-axis maximum rate, mm/min)
$111 = 3000.000    (Y-axis maximum rate, mm/min)
$112 = 500.000    (Z-axis maximum rate, mm/min)
$113 = 3000.000   
$114 = 3000.000   
$115 = 3000.000   
$120 = 120.000    (X-axis acceleration, mm/sec^2)
$121 = 120.000    (Y-axis acceleration, mm/sec^2)
$122 = 60.000    (Z-axis acceleration, mm/sec^2)
$123 = 120.000   
$124 = 120.000   
$125 = 120.000   
$130 = 300.000    (X-axis maximum travel, millimeters)
$131 = 339.000    (Y-axis maximum travel, millimeters)
$132 = 43.000    (Z-axis maximum travel, millimeters)
$133 = 300.000   
$134 = 300.000   
$135 = 300.000   
ok
>>> $G
[GC:G0 G54 G17 G21 G90 G94 M5 M9 T0 F0 S0]
ok
*** Connected to GRBL 1.3a

**Comment by aimeiz** [comment]
It seems that GRBL ESP responds for $I command in different way on COM connection and telnet connection, but most of CNC senders tolerate it. I.e. LaserGRBL, GRBLPlotter.

**Reply by breiler** [reply]
Connect to it with the GRBL firmware profile instead. I should probably remove the GRBL ESP32 firmware alternative in UGS.

Barts GRBL_ESP32 project is deprecated and you should consider upgrading to FluidNC.

**Reply by MarquinaLuz** [reply]
Una pregunta, si mi placa de desarrollo es ESP32, en UGS mi firmware es GRBL ESP 32 pero aquí menciona que está obsoleto. Debo elegir GRBL y lo va a aceptar?

**Reply by breiler** [reply]
@MarquinaLuz if you have this installed https://github.com/bdring/Grbl_Esp32 it is better to connect to it using the plain GRBL firmware profile in UGS.

**Comment by aimeiz** [comment]
I tested with laserGRBL - it doesn't send $I on connection, but when I send $I manually GRBLESP responded correctly, sending as follows:
[VER:1.3a. 20211103:]
[OPT: MHBSW]
[MSG: using machine:MKS TINYBEE]
.......
........
Sorry. LASERGRBL doesn't allow copy paste in console window, so I had to write manualy.

**Comment by aimeiz** [comment]
I tried to simulate situation using RealTerm program - telnet connection.
If I sent $I with only LF code I got same response like UGS - just ok.
when I send only with CR - I got error 3
When sent both CR and LF _ got right response:
[VER:1.3a.20211103:]
                                                          
[OPT:MPHBSW]
                                                                  
[MSG:Using machine:MKS TINYBEE]
                                               
[MSG:Mode=STA:SSID=Marcin_2.4G:Status=Connected:IP=192.168.1.153:MAC=D4-8A-FC-DF
-D3-F8]
                                                                       
[MSG:No BT]
                                                                   
ok
                                                                            
ok
                                                                            
So I suspect that UGS sends $I just with LF instead of CRLF.
Pls check and correct.

**Reply by breiler** [reply]
I've just installed the latest code from Grbl_Esp32 on main branch and then connected to it using the GRBL firmware profile without any problems using both USB and TCP/IP. I can not reproduce this problem:

```
*** Connecting to tcp://192.168.0.1:23
Grbl 1.3a ['$' for help]
<Idle|MPos:0.000,0.000,0.000|FS:0,0>
*** Fetching device status
>>> ?
<Idle|MPos:0.000,0.000,0.000|FS:0,0>
ok
>>> $I
[VER:1.3a.20211103:]
[OPT:PHBSW]
[MSG:Using machine:Test Drive - Demo Only No I/O!]
[MSG:Mode=AP:SSDI=GRBL_ESP:IP=192.168.0.1:MAC=24-62-AB-E4-4E-35]
[MSG:No BT]
ok
>>> $$
$0 = 3    (Step pulse time, microseconds)
$1 = 250    (Step idle delay, milliseconds)
$2 = 0    (Step pulse invert, mask)
$3 = 0    (Step direction invert, mask)
$4 = 0    (Invert step enable pin, boolean)
$5 = 1    (Invert limit pins, boolean)
$6 = 0    (Invert probe pin, boolean)
$10 = 1    (Status report options, mask)
$11 = 0.010    (Junction deviation, millimeters)
$12 = 0.002    (Arc tolerance, millimeters)
$13 = 0    (Report in inches, boolean)
$20 = 0    (Soft limits enable, boolean)
$21 = 0    (Hard limits enable, boolean)
$22 = 0    (Homing cycle enable, boolean)
$23 = 3    (Homing direction invert, mask)
$24 = 200.000    (Homing locate feed rate, mm/min)
$25 = 2000.000    (Homing search seek rate, mm/min)
<Idle|MPos:0.000,0.000,0.000|FS:0,0|WCO:0.000,0.000,0.000>
$26 = 250.000    (Homing switch debounce delay, milliseconds)
$27 = 1.000    (Homing switch pull-off distance, millimeters)
$30 = 1000.000    (Maximum spindle speed, RPM)
$31 = 0.000    (Minimum spindle speed, RPM)
$32 = 0    (Laser-mode enable, boolean)
$100 = 100.000    (X-axis travel resolution, step/mm)
$101 = 100.000    (Y-axis travel resolution, step/mm)
$102 = 100.000    (Z-axis travel resolution, step/mm)
$103 = 100.000   
$104 = 100.000   
$105 = 100.000   
$110 = 1000.000    (X-axis maximum rate, mm/min)
$111 = 1000.000    (Y-axis maximum rate, mm/min)
$112 = 1000.000    (Z-axis maximum rate, mm/min)
$113 = 1000.000   
$114 = 1000.000   
$115 = 1000.000   
$120 = 200.000    (X-axis acceleration, mm/sec^2)
$121 = 200.000    (Y-axis acceleration, mm/sec^2)
$122 = 200.000    (Z-axis acceleration, mm/sec^2)
$123 = 200.000   
$124 = 200.000   
$125 = 200.000   
$130 = 300.000    (X-axis maximum travel, millimeters)
$131 = 300.000    (Y-axis maximum travel, millimeters)
$132 = 300.000    (Z-axis maximum travel, millimeters)
$133 = 300.000   
$134 = 300.000   
$135 = 300.000   
ok
>>> $G
[GC:G0 G54 G17 G21 G90 G94 M5 M9 T0 F0 S0]
ok
*** Connected to GRBL 1.3a
```

When using the Grbl_Esp32 firmware profile I can't get it to work. I will remove that profile in future releases of UGS.

**Reply by aimeiz** [reply]
So it my be just GRBL profile to be used instead of Grbl_ESP32? It didn't work as well, but I gonna chect again and maybe try to reinstall firmware on my cnc machine and let You know. It could depend on GRBL machine setup, however i do not see dependence.

**Reply by aimeiz** [reply]
I checked again with UGS GRBL profile, but same result:
>>> $I
ok
*** Could not detect the GRBL version
**** Connection closed ****
No response on $I command, but from LaserGRBL and also just simple terminal program it responds correctly if CRLF is send after $I

**Reply by breiler** [reply]
Then this is an issue with GRBL_ESP32, maybe how it is compiled on your specific environment since I am running the same version. I can't reproduce it, therefore there is nothing I can fix.

And I am reluctant to add `CR` to line endings in the protocol as a workaround for your specific problem. This will effect the controller communication for all users - which has worked for thousands (maybe millions?) of machines over a decade.

The developers of GRBL_ESP32 project have moved on to FluidNC and you will have to switch to that to get any help from those developers or use another sender software.

**Reply by aimeiz** [reply]
Hi.
I checked with my previous machine controller from MKS and also with my old XY plotter running version 1.1 - same results.
Firmware for MKS-DLC32 board is on separate account in github:
https://github.com/makerbase-mks/MKS-DLC32-FIRMWARE/tree/main
Maybe with this one You will be able to replicate problem and solve it.

**Reply by breiler** [reply]
I was able to reproduce the problem with the MKS-firmware. It seems that when running in WiFi mode it needs a short pause before querying the build information. A two second pause worked for me.

It is available in the nightly build here:
https://github.com/winder/Universal-G-Code-Sender?tab=readme-ov-file#downloads

**Comment by aimeiz** [comment]
I understand Your approach. I studied UGS code and it always use "\n", so maybe, as You suggested problem is in GRBLESP.
Regard to testing - I am using STA, so machione connect to my LAN, get IP... I see You tested in AP mode, gut it shouldn't make any difference.

If I knew how to compile UGS with Visual Studio, I could experiment modyfying:
GetFirmwareVersionCommand.java lines 33 - 35
    public GetFirmwareVersionCommand() {
        super("$I");
    }

to super("$I\r");

I tried to find definition of super() function, but couldn't. I am not experienced in java projects.

**Comment by aimeiz** [comment]
I tried to do Your test. I cloned GEBLESP32 repository from link I gave You to Visual Studio Code / PlatfotmIO as is, without any customization. Compiled and uploaded it to Wemos ESP32 board.
Then tried to connect it with UGS and got the same problem.
No problems with LaserGRBL and GRBLPlotter programs.
Interesting how You built it and uploaded which board to to make it working?

**Reply by breiler** [reply]
Did you try the latest nightly build?

https://github.com/winder/Universal-G-Code-Sender/discussions/2526#discussioncomment-10213862

**Comment by aimeiz** [comment]
I tried 2 days ago, but maybe You put new version. Delay on WiFi makes sense for me, since net communication could create some surprises.

**Reply by breiler** [reply]
I pushed these changes two hours ago.

**Comment by aimeiz** [comment]
The most recent Nightbuild (SNAPSHOT) gives same result:
*** Connecting to tcp://192.168.1.153:23
Grbl 1.3a ['$' for help]
<Alarm|WPos:-158.899,-83.299,35.215,0.000,0.000|Bf:15,1200|FS:0,0>
*** Fetching device status
>>> ?
<Alarm|WPos:-158.899,-83.299,35.215,0.000,0.000|Bf:15,1199|FS:0,0|WCO:158.899,83.299,-35.215,0.000,0.000>
>>> 
ok
>>> $I
ok
*** Could not detect the GRBL version
**** Connection closed ****

I tested my 4 machines - same results.

**Comment by aimeiz** [comment]
By the way how to build UGS code? I tried to clone with Visual Studio and Visual Studio code, but Perhaps I need some plugins to build such java project.

**Comment by aimeiz** [comment]
OK I found instruction how to build. It needs another IDE - Apache Maven with JDK Java17.

**Comment by aimeiz** [comment]
I downloaded Nightbuild from this link https://github.com/winder/Universal-G-Code-Sender/releases/download/nightly/win64-ugs-platform-app-2.0-SNAPSHOT.zip it seems to be recent exe created today at 15:00, but is older than 2 hours ago :(

**Comment by aimeiz** [comment]
Ok I erased previous download and loaded again. still exe file is from 15:00, but it works now both GRBL and GRBL profiles.
*** Connecting to tcp://192.168.1.153:23
Grbl 1.3a ['$' for help]
<Alarm|WPos:-158.899,-83.299,35.215,0.000,0.000|Bf:15,1200|FS:0,0>
*** Fetching device status
>>> ?
<Alarm|WPos:-158.899,-83.299,35.215,0.000,0.000|Bf:15,1199|FS:0,0>
>>> 
ok
ok
>>> $I
[VER:1.3a.20211103:]
[OPT:MPHBSW]
[MSG:Using machine:MKS TINYBEE]
[MSG:Mode=STA:SSID=Marcin_2.4G:Status=Connected:IP=192.168.1.153:MAC=D4-8A-FC-DF-D3-F8]
[MSG:No BT]
ok
>>> $$
<Alarm|WPos:-158.899,-83.299,35.215,0.000,0.000|Bf:15,1200|FS:0,0>
$0 = 10    (Step pulse time, microseconds)
$1 = 254    (Step idle delay, milliseconds)
$2 = 0    (Step pulse invert, mask)
$3 = 5    (Step direction invert, mask)
$4 = 0    (Invert step enable pin, boolean)
$5 = 1    (Invert limit pins, boolean)
$6 = 1    (Invert probe pin, boolean)
$10 = 2    (Status report options, mask)
$11 = 0.010    (Junction deviation, millimeters)
$12 = 0.002    (Arc tolerance, millimeters)
$13 = 0    (Report in inches, boolean)
$20 = 1    (Soft limits enable, boolean)
$21 = 0    (Hard limits enable, boolean)
$22 = 1    (Homing cycle enable, boolean)
$23 = 3    (Homing direction invert, mask)
$24 = 250.000    (Homing locate feed rate, mm/min)
$25 = 1000.000    (Homing search seek rate, mm/min)
$26 = 250.000    (Homing switch debounce delay, milliseconds)
$27 = 3.000    (Homing switch pull-off distance, millimeters)
$30 = 10000.000    (Maximum spindle speed, RPM)
$31 = 0.000    (Minimum spindle speed, RPM)
$32 = 1    (Laser-mode enable, boolean)
$100 = 801.200    (X-axis travel resolution, step/mm)
$101 = 802.500    (Y-axis travel resolution, step/mm)
$102 = 802.500    (Z-axis travel resolution, step/mm)
$103 = 802.500   
$104 = 802.500   
$105 = 100.000   
$110 = 3000.000    (X-axis maximum rate, mm/min)
$111 = 3000.000    (Y-axis maximum rate, mm/min)
$112 = 500.000    (Z-axis maximum rate, mm/min)
$113 = 3000.000   
$114 = 3000.000   
$115 = 3000.000   
$120 = 120.000    (X-axis acceleration, mm/sec^2)
$121 = 120.000    (Y-axis acceleration, mm/sec^2)
$122 = 60.000    (Z-axis acceleration, mm/sec^2)
$123 = 120.000   
$124 = 120.000   
$125 = 120.000   
$130 = 300.000    (X-axis maximum travel, millimeters)
$131 = 339.000    (Y-axis maximum travel, millimeters)
$132 = 43.000    (Z-axis maximum travel, millimeters)
$133 = 300.000   
$134 = 300.000   
$135 = 300.000   
ok
>>> $G
[GC:G0 G54 G17 G21 G90 G94 M5 M9 T0 F0 S0]
ok
*** Connected to GRBL 1.3a

So success !!! - Thank You Joacim. Greetings to Sweden :).
